tests/late_joiners.py

```diff
@@ -0,0 +1,216 @@
+                clients.append(es.enter_context(first_node.user_client(format="json")))
+                for backup in backups:
+                    clients.append(es.enter_context(backup.user_client(format="json")))
+                node_id = 0
+                for id in range(1001, (1001 + TOTAL_REQUESTS)):
+                    node_id += 1
+                    c = clients[node_id % len(clients)]
+                    try:
+                        resp = c.rpc("LOG_record", {"id": id, "msg": long_msg})
+                    except Exception:
+                        LOG.info("Trying to access a suspended node")
+                    try:
+                        cur_primary, cur_term = network.find_primary()
+                        term_info[cur_term] = cur_primary.node_id
+                    except Exception:
+                        LOG.info("Trying to access a suspended node")
+                    id += 1
+
+                # wait for the last request to commit
+                second_catchup_msg = "Hello world Hello!"
+                check_commit(
+                    c.rpc("LOG_record", {"id": 2000, "msg": second_catchup_msg}),
+                    result=True,
+                )
+                check(
+                    c.rpc("LOG_get", {"id": 2000}), result={"msg": second_catchup_msg},
+                )
+
+            with last_node.user_client(format="json") as c:
+                check(
+                    c.rpc("LOG_get", {"id": 1000}), result={"msg": first_catchup_msg},
+                )
+            with last_node.user_client(format="json") as c:
+                check(
+                    c.rpc("LOG_get", {"id": 2000}), result={"msg": second_catchup_msg},
+                )
+
+            # replace the 2 backups with the 2 new nodes, kill the old ones and ensure we are still making progress
+            for node in nodes_to_kill:
+                LOG.info(f"Stopping node {node.node_id}")
+                node.stop()
+
+            for i, node in enumerate(nodes_to_keep):
+                with node.user_client(format="json") as c:
+                    final_msg = "Goodby world!"
+                    check_commit(
+                        c.rpc("LOG_record", {"id": 3000 + i, "msg": final_msg}),
+                        result=True,
+                    )
+                    check(
+                        c.rpc("LOG_get", {"id": 3000 + i}), result={"msg": final_msg},
+                    )
+
+                # all the nodes should be caught up by now
+
+                if not args.skip_suspension:
+                    # assert that view changes actually did occur
+                    assert len(term_info) > 1
+
+                    LOG.success("----------- terms and primaries recorded -----------")
+                    for term, primary in term_info.items():
+                        LOG.success(f"term {term} - primary {primary}")
+
+
+if __name__ == "__main__":
+
+    def add(parser):
+        parser.add_argument(
```

Comment:
It seems that `skip-suspension` is set for the CI runs. Should this option disappear so that we test the full behaviour in the CI?

Reply:
We will activate suspension when we implement rollback (next PR) so that view changes can happen